### PR TITLE
FlowDistribution

### PR DESCRIPTION
**Pull request by botev**
Following the conversation in #70

### CODE REVIEW

**Comment by botev** [comment]
Currently, I have only modified the `planar_flow` to be single transformation and to return directly the `log_det` of the transformation. I'm not too sure if the way I've used variable scopes is currently correct, so any suggestions on that would be nice. I need to also to change accordingly and the other flows. I've actually have implemented before quite a few so would be happy to add more. 

Additionally, discussion on the weather you guys like the `FlowDistribution` abstraction is needed. Tests for that have to be added as well.

**Comment by thjashin** [comment]
Cool, I will start looking at this once I finished the matrix normal one.

**Comment by thjashin** [comment]
Back from rushing an icml paper. Sorry for being late but I'm reading the code now.

**Comment by botev** [comment]
Sure no worries, hope ICML went well.

File: examples/normalizing_flows/dlgm_nf.py [hunk 1]
```diff
@@ -29,14 +30,17 @@ def vae(observed, n, x_dim, z_dim, n_particles):
...
+        return zs.repeated_flow(zs.planar_normalizing_flow, samples, n_iters=n_flows)
+
     with zs.BayesianNet() as variational:
         lz_x = tf.layers.dense(tf.to_float(x), 500, activation=tf.nn.relu)
         lz_x = tf.layers.dense(lz_x, 500, activation=tf.nn.relu)
         z_mean = tf.layers.dense(lz_x, z_dim)
         z_logstd = tf.layers.dense(lz_x, z_dim)
-        z = zs.Normal('z', z_mean, logstd=z_logstd, group_ndims=1,
-                      n_samples=n_particles)
+        z = zs.NormalFlow('z', forward, mean=z_mean, logstd=z_logstd, group_ndims=1,
```
Comment (thjashin) [inline examples/normalizing_flows/dlgm_nf.py]: I feel `NormalFlow` is a bit too restricted. And it also makes the internal Normal node invisible. I'd prefer sth. like
```python
z_pre = zs.Normal('z_pre', ...)
z = zs.Flow('z', flow=zs.planar_normalizing_flow, n_iters=n_flows)
```

File: zhusuan/distributions/special.py [hunk 2]
```diff
@@ -157,3 +158,77 @@ def _prob(self, given):
...
+
+class FlowDistribution(Distribution):
+    """
+    The class of FlowDistribution distribution.
+    The distribution describes variable which is sampled from a base
+    distribution and then is passed through an invertible function.
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    :param name: A string. The name of the `StochasticTensor`. Must be unique
+        in the `BayesianNet` context.
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: The signature is incorrect

File: zhusuan/distributions/special.py [hunk 3]
```diff
@@ -157,3 +158,77 @@ def _prob(self, given):
...
+    :param base: An instance of `Distribution` parametrizing the base distribution.
+    :param forward: A forward function which describes how we transform the samples
+        from the base distribution. The signature of the function should be:
+            transformed, log_det = forward(base_samples)
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: We should restrict how the forward function works. Does it apply transformation on each value (of `value_shape`) individually, or does it apply transformation on the group of values (of shape `batch_shape[-group_ndims:] + value_shape`)? There needs some careful design.

File: zhusuan/distributions/special.py [hunk 4]
```diff
@@ -157,3 +158,77 @@ def _prob(self, given):
...
+            is_continuous=base.dtype.is_floating,
+            group_ndims=group_ndims,
+            is_reparameterized=False,
+            **kwargs)
+
+    def _value_shape(self):
+        return self.base.value_shape()
+
+    def _get_value_shape(self):
+        return self.base.get_value_shape()
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: As stated above, the new value shape is not always the same as the original one, e.g., in the VAE case.

File: zhusuan/distributions/special.py [hunk 5]
```diff
@@ -157,3 +158,77 @@ def _prob(self, given):
...
+
+    def _batch_shape(self):
+        return self.base.batch_shape()
+
+    def _get_batch_shape(self):
+        return self.base.get_batch_shape()
+
+    def _sample(self, n_samples):
+        return self.sample_and_log_prob(n_samples)[0]
```
Comment (thjashin) [inline zhusuan/distributions/special.py]: This seems to not follow our discussion [here](https://github.com/thu-ml/zhusuan/issues/70#issuecomment-345773647)? Also the `StochasticTensor` part.

File: zhusuan/model/base.py [hunk 6]
```diff
@@ -115,9 +115,16 @@ def tensor(self):
...
-                self._tensor = self.sample(self._n_samples)
+                self._tensor, self._local_log_prob = self.sample_and_log_prob(self._n_samples)
         return self._tensor
 
+    @property
+    def local_log_prob(self):
+        tensor = self.tensor
+        if not hasattr(self, '_local_log_prob'):
+            self._local_log_prob = self.log_prob(tensor)
+        return self._local_log_prob
```
Comment (thjashin) [inline zhusuan/model/base.py]: Nice job. Meanwhile can you remove the `log_prob(given)` function in `StochasticTensor`? We should make `StochasticTensor` more static and leave these functions to `StochasticTensor.distribution.log_prob(given)`

File: zhusuan/transform.py [hunk 7]
```diff
@@ -10,10 +10,94 @@
 
 
 __all__ = [
+    'repeated_flow',
     'planar_normalizing_flow',
 ]
 
 
+def repeated_flow(base_flow_function, samples, n_iters, scope=None,
```
Comment (thjashin) [inline zhusuan/transform.py]: Nice refactor. Thanks for the improvements.

File: examples/normalizing_flows/dlgm_nf.py (hunk 1 above)
Comment (botev) [reply examples/normalizing_flows/dlgm_nf.py]: However, this would require some extra internal machinery or require that the input to the flow is an instance of `StochastcTensor` in order to still be able to use its sampling and log-probability methods

File: zhusuan/distributions/special.py (hunk 3 above)
Comment (botev) [reply zhusuan/distributions/special.py]: So, I think that the function signature should be on each value independently, but to work with a batch of those as inputs. Since the flow distribution is defined as a transformation of each variable it is a deterministic function in each value independently. As long as the function can work with a batch I think its fine. Potentially we might want to flatten all the batch dimensions when it is inputted to forward and then reshaping back to make it easier for the implementators to not have the option of multiple batch dimensions.

File: zhusuan/distributions/special.py (hunk 5 above)
Comment (botev) [reply zhusuan/distributions/special.py]: I think I've decided to leave the `_sample` be implemented as I circumvented the issue with tensor using the `local_log_prob` property. If you still want I can make it not implemented, but I'm not sure it is needed.

File: zhusuan/transform.py (hunk 7 above)
Comment (botev) [reply zhusuan/transform.py]: After we sort out correctly all of the flow distribution details, I was intending to add at least a couple of more of the flows in the literature as well.
